src/libraries/System.Private.CoreLib/src/System/Globalization/Normalization.Unix.cs

```diff
@@ -41,26 +42,52 @@ internal static string Normalize(string strInput, NormalizationForm normalizatio
 
             ValidateArguments(strInput, normalizationForm);
 
-            char[] buf = new char[strInput.Length];
-
-            for (int attempts = 2; attempts > 0; attempts--)
+            char[]? toReturn = null;
+            try
             {
-                int realLen = Interop.Globalization.NormalizeString(normalizationForm, strInput, strInput.Length, buf, buf.Length);
+                Span<char> buffer = strInput.Length <= 512
+                    ? stackalloc char[512]
+                    : (toReturn = ArrayPool<char>.Shared.Rent(strInput.Length));
 
-                if (realLen == -1)
+                for (int attempt = 0; attempt < 2; attempt++)
                 {
-                    throw new ArgumentException(SR.Argument_InvalidCharSequenceNoIndex, nameof(strInput));
+                    realLen = Interop.Globalization.NormalizeString(normalizationForm, strInput, strInput.Length, buffer);
+
+                    if (realLen == -1)
+                    {
+                        throw new ArgumentException(SR.Argument_InvalidCharSequenceNoIndex, nameof(strInput));
+                    }
+
+                    if (realLen <= buffer.Length)
+                    {
+                        return new string(buffer.Slice(0, realLen));
+                    }
+
+                    Debug.Assert(realLen > 512);
+
+                    if (attempt == 0)
+                    {
+                        if (toReturn != null)
+                        {
+                            // Clear toReturn first to ensure we don't return the same buffer twice
+                            char[] temp = toReturn;
+                            toReturn = null;
+                            ArrayPool<char>.Shared.Return(temp);
+                        }
+
+                        buffer = toReturn = ArrayPool<char>.Shared.Rent(realLen);
+                    }
                 }
 
-                if (realLen <= buf.Length)
+                throw new ArgumentException(SR.Argument_InvalidCharSequenceNoIndex, nameof(strInput));
+            }
+            finally
+            {
+                if (toReturn != null)
                 {
-                    return new string(buf, 0, realLen);
+                    ArrayPool<char>.Shared.Return(toReturn);
```

Comment:
Heads up: we generally don't return buffers to the shared pool inside finally blocks. The sole exception is where we're in 100% control of every single code path that might be invoked inside the preceding try.

Reply:
> Heads up: we generally don't return buffers to the shared pool inside finally blocks. The sole exception is where we're in 100% control of every single code path that might be invoked inside the preceding try.

We're inconsistent on this, e.g.
https://github.com/dotnet/runtime/blob/26b6e4ea97a627ab800362b2c10f32ebecea041d/src/libraries/System.Private.CoreLib/src/System/IO/Stream.cs#L114-L130

Reply:
I know, but we should try to get it under check for new code (like this). The GitHub UI is misbehaving for me so I can't see the rest of the code. So this usage might not be problematic at all but somebody should double check.